napari/utils/colormaps/colormap_utils.py

```diff
@@ -618,7 +618,27 @@ def ensure_colormap(colormap: ValidColormapArg) -> Colormap:
     """
     with AVAILABLE_COLORMAPS_LOCK:
         if isinstance(colormap, str):
-            name = colormap
+            # if the name starts with a '#' we assume it to be a hex color
+            # we then check if the hex color is one of r,g,b,c,m,y,w
+            # in which case we can reuse the associated colormap.
+            if colormap.startswith('#') and len(colormap) in [4, 5, 7, 9]:
+                name = _hexcolor_to_name(colormap)
+                if name.startswith('custom'):
+                    # generate a linear colormap from black to that color
+                    # and add it to the dictionary of available colormaps
+                    cmap = _colormap_from_colors(colormap, name)
+                    if cmap is not None:
+                        AVAILABLE_COLORMAPS[name] = cmap
+                    else:
+                        raise ValueError(
+                            trans._(
+                                "Unable to convert the hex color string into a colormap",
+                                deferred=True,
+                            )
+                        )
+            else:
+                name = colormap
```

Comment:
I think that you should use here `vispy.color.Color` to recognize any string that is a valid vispy color. Maybe something like this (plus handle vispy_or_mpl_colormap).

```suggestion
            cmap = AVAILABLE_COLORMAPS.get(colormap, AVAILABLE_COLORMAPS.get(f"custom-{colormap}", None)
            if cmap is None:
                cmap = _colormap_from_colors(colormap, name=f"custom-{colormap}")
                for cmap_ in AVAILABLE_COLORMAPS.values():
                    if cmap_.controls == cmap.controls and cmap_.colors == cmap.colors and cmap_.interpolation == cmap.interpolation:
                        cmap = cmap_
                        break
            name = cmap.name
            AVAILABLE_COLORMAPS[name] = cmap
```

Reply:
@Czaki thank you very much for the idea, indeed this makes my _hexcolor_to_name() function completely unnecessary as you are directly testing the colors of the newly generated cmap against any of the existing ones.

There may be a slight issue with the way the `.colors` attribute is defined as a ColorArray of a single nx4 numpy array rather than a list (?) of colors (like [in the vispy example given for ColorArray](https://www.ssec.wisc.edu/~davidh/tmp/vispy/api/vispy.color.color_array.html#vispy.color.color_array.ColorArray), which messes up your `cmap_.colors == cmap.colors` test, in the sense that an array of bools, one for each component is returned rather than a single true/false value.

So instead this worked for me (testing shapes are equal and using `np.all()`):
```python
                for cmap_ in AVAILABLE_COLORMAPS.values():
                    if np.array_equal(cmap_.controls, cmap.controls) and cmap_.colors.shape == cmap.colors.shape and np.all(cmap_.colors == cmap.colors) and cmap_.interpolation == cmap.interpolation:
                        cmap = cmap_
                        break
```

And here's how I've tested it so far:
```python
import napari.utils.colormaps.colormap_utils as cu
cmap = cu.ensure_colormap('#f0fF')
cu.AVAILABLE_COLORMAPS.get(cmap.name).name
```
returns `'magenta'`

I will check this some more and amend the code as necessary.

Cheers,
Egor